### ISSUE TITLE
Harmonic slow to evaluate at machine precision

### ISSUE THREAD

**Issue by nielsw2** [BlackHolePerturbationToolkit/SpinWeightedSpheroidalHarmonics#2]
It seems the harmonic is much slower to evaluate at machine precision than the higher precision version. I suspect this a difference between the Eigenvalue and Leaver methods.

S = SpinWeightedSpheroidalHarmonicS[1, 4, 4, 0.7`20]
S[\[Pi]/3, 0.5`20] // Timing 
 {0.000106, -0.15909486841868710 + 0.34762862953517741 I}

vs

S = SpinWeightedSpheroidalHarmonicS[1, 4, 4, 0.7]
S[\[Pi]/3, 0.5] // Timing
S[\[Pi]/3., 0.5] // Timing

{0.198584, -0.0482024 + 0.105324 I}
{0.067477, -0.0482024 + 0.105324 I}

Note it is much quicker to evaluate when the theta is not exact also.

**Comment by barrywardell** [comment]
Yes, it looks like it's the difference between Eigenvalue and Leaver methods, rather than between machine and higher precision. The following expanded version of your example demonstrates this quite clearly.

```Mathematica
In[1]:= (S1 = SpinWeightedSpheroidalHarmonicS[1, 4, 4, 0.7`20]) // Timing
Out[1]= {0.012664, SpinWeightedSpheroidalHarmonicSFunction[1, 4, 4, 0.70000000000000000000, << 23>>, Method -> "Leaver"]}

In[2]:= (S1E = SpinWeightedSpheroidalHarmonicS[1, 4, 4, 0.7`20, Method -> "Eigenvalue"]) // Timing
Out[2]= {0.367774, SpinWeightedSpheroidalHarmonicSFunction[1, 4, 4, 0.70000000000000000000, <<54>>, Method -> "Eigenvalue"]}

In[3]:= (S2 = SpinWeightedSpheroidalHarmonicS[1, 4, 4, 0.7]) // Timing
Out[3]= {0.013856, SpinWeightedSpheroidalHarmonicSFunction[1, 4, 4, 0.7, <<54>>, Method -> "Eigenvalue"]}

In[4]:= S1[\[Pi]/3, 0.5`20] // Timing
Out[4]= {0.000236, -0.04820237132041662 + 0.10532410283884410 I}

In[5]:= S1[\[Pi]/3.`20, 0.5`20] // Timing
Out[51]= {0.000129, -0.04820237132041662 + 0.10532410283884410 I}

In[6]:= S1E[\[Pi]/3, 0.5`20] // Timing
Out[52]= {0.046123, -0.04820237132060151319 + 0.10532410283924809499 I}

In[7]:= S1E[\[Pi]/3.`20, 0.5`20] // Timing
Out[53]= {0.024091, -0.04820237132060151319 + 0.1053241028392480950 I}

In[8]:= S2[\[Pi]/3, 0.5] // Timing
Out[54]= {0.048848, -0.0482024 + 0.105324 I}

In[9]:= S2[\[Pi]/3., 0.5] // Timing
Out[55]= {0.016065, -0.0482024 + 0.105324 I}
```

**Comment by barrywardell** [comment]
One possible solution is to always use Leaver's method by default for computing the harmonic. The "Spectral" method is useful to provide an initial seed for the Leaver eigenvalue calculation, but is not so clearly useful for computing the harmonic. The only reason to keep using the Spectral method for the harmonic is that it might be more robust in some cases, but pull request #19 fixes most of these.

**Comment by barrywardell** [comment]
This was resolved in f272e84b by making the Leaver method the default also for machine precision. The SphericalExpansion method was also significantly sped up in f6d6f9f by reducing the number of coupling terms used.
